commands/url_indexer.py

```diff
@@ -0,0 +1,181 @@
+import sublime, sublime_plugin
+
+from collections import namedtuple
+from pathlib import Path
+from glob import iglob
+import json
+import subprocess
+
+from .utils.common import Logger
+
+logger = Logger(__name__)
+
+IndexEntry = namedtuple("IndexEntry", ["full_method", "url", "method_name", "kwargs"])
+
+
+class URLIndex(dict):
+    @classmethod
+    def init_with_index(cls, index):
+        self = cls()
+        self = {i[2].lower(): IndexEntry(*i) for i in index if i[2] is not None}
+        return self
+
+
+class DjangoUrlsIndexer(sublime_plugin.WindowCommand):
+    """A window command to index django url files."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._is_running = False
+
+        # urls.py path to url names
+        self.window_index = {}
+        self.python_interpreter = None
+
+    def clear_index(self):
+        self.index = {}
+
+    def set_window_index(self, window, index):
+        self.window_index[window] = URLIndex.init_with_index(index)
+
+    def index_project_folders(self, window, project_folders):
+        if self._is_running:
+            return
+        self._is_running = True
+
+        logger.debug("Indexing project data...")
+        managepy = None
+        for folder in project_folders:
+            path_str = folder.get("path")
+            if not path_str:
+                continue
+            path = Path(path_str)
+            exclude = folder.get("folder_exclude_patterns")
+            managepy = path.glob("**/manage.py")
+            if not managepy:
+                continue
+
+        if not managepy:
+            logger.info("No manage.py found in project.  URL Index stopped.")
+            return
+
+        managepy = next(managepy)
+
+        this_path = Path(__file__).parent.absolute() / "utils"
+        cmd = f'import sys; sys.path.append("{this_path}"); from url_extract import run; run()'
+        result = subprocess.run(
+            [self.python_interpreter, managepy, "shell", "-c", cmd],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        if result.stderr:
+            logger.error(result.stderr)
+
+        if not result.stdout:
+            logger.warning("No output from url_extract script.")
+            return
+
+        self.set_window_index(window, json.loads(result.stdout))
+
+        logger.debug("Indexing complete.")
+
+        self._is_running = False
+
+    def get_index_for_window(self, window):
+        return self.window_index.get(window)
+
+
+djurls_indexer = DjangoUrlsIndexer(sublime.active_window())
+
+
+class DjangoUrlsPlugin(sublime_plugin.EventListener):
+    # def on_activated(self, view):
```

Comment:
Do we need this?

Reply:
It's helpful to quickly auto-refresh the index when debugging, but I'll remove it.